src/operations/blas1/WGAtomicReduction.hpp

```diff
@@ -0,0 +1,136 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  portBLAS: BLAS implementation using SYCL
+ *
+ *  @filename WGAtomicReduction.hpp
+ *
+ **************************************************************************/
+
+#ifndef WG_ATOMIC_REDUCTION_HPP
+#define WG_ATOMIC_REDUCTION_HPP
+#include "operations/blas1_trees.h"
+#include "operations/blas_operators.hpp"
+
+namespace blas {
+
+/*! WGAtomicReduction.
+ * @brief This class implement a device size reduction using all WG to compute
+ * and atomics operation to combine the results.
+ *
+ * */
+template <typename operator_t, typename lhs_t, typename rhs_t>
+WGAtomicReduction<operator_t, lhs_t, rhs_t>::WGAtomicReduction(lhs_t& _l,
+                                                               rhs_t& _r)
+    : lhs_(_l), rhs_(_r){};
+
+template <typename operator_t, typename lhs_t, typename rhs_t>
+PORTBLAS_INLINE typename WGAtomicReduction<operator_t, lhs_t, rhs_t>::index_t
+WGAtomicReduction<operator_t, lhs_t, rhs_t>::get_size() const {
+  return rhs_.get_size();
+}
+
+template <typename operator_t, typename lhs_t, typename rhs_t>
+PORTBLAS_INLINE bool WGAtomicReduction<operator_t, lhs_t, rhs_t>::valid_thread(
+    cl::sycl::nd_item<1> ndItem) const {
+  return true;
+}
+
+template <typename operator_t, typename lhs_t, typename rhs_t>
+PORTBLAS_INLINE typename WGAtomicReduction<operator_t, lhs_t, rhs_t>::value_t
+WGAtomicReduction<operator_t, lhs_t, rhs_t>::eval(cl::sycl::nd_item<1> ndItem) {
+  auto atomic_res = sycl::atomic_ref<value_t, sycl::memory_order::relaxed,
+                                     sycl::memory_scope::device,
+                                     sycl::access::address_space::global_space>(
+      lhs_.get_data()[0]);
+  const auto size = rhs_.get_size();
+  int lid = ndItem.get_global_linear_id();
+  value_t val = operator_t::template init<rhs_t>();
+
+  // First loop for big arrays
+  for (int id = lid; id < size;
+       id += ndItem.get_local_range()[0] * ndItem.get_group_range()[0]) {
```

Comment:
Otherwise I think it would increase readability & potentially reduce the small multiplication overhead to have that kind-of _striding_ value fixed at `WGAtomicReduction` instantiation (as a const static member attribute for eg) since it's constant and doesn't depend on present item.

Reply:
1) I did not know I could access that value directly, thank you @OuadiElfarouki. 
2) I agree with you that storing the striding value adds readability and it could increase performance, I will update this kernel.